pubsub/google/cloud/pubsub_v1/publisher/batch/base.py

```diff
@@ -0,0 +1,130 @@
+    users may need to implement something based on a different concurrency
+    model.
+
+    This class defines the interface for the Batch implementation;
+    subclasses may be passed as the ``batch_class`` argument to
+    :class:`~.pubsub_v1.client.PublisherClient`.
+    """
+    @property
+    @abc.abstractmethod
+    def client(self):
+        """Return the client used to create this batch.
+
+        Returns:
+            ~.pubsub_v1.client.PublisherClient: A publisher client.
+        """
+        raise NotImplementedError
+
+    @property
+    @abc.abstractmethod
+    def settings(self):
+        """Return the settings for this batch.
+
+        Returns:
+            ~.pubsub_v1.types.Batching: The settings for batch
+                publishing. These should be considered immutable once the batch
+                has been opened.
+        """
+        raise NotImplementedError
+
+    @property
+    @abc.abstractmethod
+    def status(self):
+        """Return the status of this batch.
+
+        Returns:
+            str: The status of this batch. All statuses are human-readable,
+                all-lowercase strings, and represented in the
+                :class:`BaseBatch.Status` enum.
+        """
+        raise NotImplementedError
+
+    @abc.abstractmethod
+    def commit(self):
+        """Asychronously commit everything in this batch.
+
+        Subclasses must define this as an asychronous method; it may be called
+        from the primary process by :meth:`check_limits`.
+        """
+        raise NotImplementedError
+
+    @abc.abstractmethod
+    def publish(self, data, **attrs):
+        """Publish a single message.
+
+        .. note::
+            Messages in Pub/Sub are blobs of bytes. They are *binary* data,
+            not text. You must send data as a bytestring
+            (``bytes`` in Python 3; ``str`` in Python 2), and this library
+            will raise an exception if you send a text string.
+
+            The reason that this is so important (and why we do not try to
+            coerce for you) is because Pub/Sub is also platform independent
+            and there is no way to know how to decode messages properly on
+            the other side; therefore, encoding and decoding is a required
+            exercise for the developer.
+
+        Add the given message to this object; this will cause it to be
+        published once the batch either has enough messages or a sufficient
+        period of time has elapsed.
+
+        Args:
+            data (bytes): A bytestring representing the message body. This
+                must be a bytestring (a text string will raise TypeError).
+            attrs (Mapping[str, str]): A dictionary of attributes to be
+                sent as metadata. (These may be text strings or byte strings.)
+
+        Raises:
+            TypeError: If the ``data`` sent is not a bytestring, or if the
+                ``attrs`` are not either a ``str`` or ``bytes``.
```

Comment:
> By the same token: Why not require the attribute keys/values to be "text strings", i.e. sequence of unicode code points?

I am doing that.

> Actually, to the extent that we're just blindly propagating them into the proto library, the real question is how the protobuf generated code handles unicode/str-style conversions, and whether we want to impose stricter semantics in the name of interplatform operability / uniqueness.

> The flip side is of course that if we do require strictly inputs of `six.text_type`, Python 2 users no longer have automagic support for publishing ascii raw strings (as valid ascii is valid utf-8 by definition). Open question as to whether this trade is worthwhile.

I think it is pretty reasonable to tell Python 2 users that they have to send `unicode` objects. The issue here is really realizing what you are sending and why. The fact that an ASCII-only binary string would _work_ is true, but part of the point that has been driven home to me by the CPS team is that users need to be explicitly aware of what they are sending. Being stringent on typing does this.

> At minimum, the comment ("These may be text strings or byte strings") is incorrect. While they are conceptually text strings, we will accept either the string or the UTF-8 encoded bytes thereof. Critically, this is not arbitrary binary (unlike the payload).

Fair point.